x/bridging_fee/ibc_middleware.go

```diff
@@ -0,0 +1,112 @@
+	return &BridgingFeeMiddleware{
+		IBCModule:        transfer,
+		ICS4Wrapper:      channelKeeper,
+		delayedAckKeeper: keeper,
+		transferKeeper:   transferKeeper,
+		feeModuleAddr:    feeModuleAddr,
+	}
+}
+
+func (im BridgingFeeMiddleware) GetBridgingFee(ctx sdk.Context) sdk.Dec {
+	return im.delayedAckKeeper.BridgingFee(ctx)
+}
+
+// Get the bridging fee param
+func (im BridgingFeeMiddleware) GetFeeRecipient(ctx sdk.Context) sdk.AccAddress {
+	return im.feeModuleAddr
+}
+
+func (im *BridgingFeeMiddleware) OnRecvPacket(ctx sdk.Context, packet channeltypes.Packet, relayer sdk.AccAddress) exported.Acknowledgement {
+	// skip check if source chain is not a rollapp
+	if !im.delayedAckKeeper.IsRollappsEnabled(ctx) {
+		return im.IBCModule.OnRecvPacket(ctx, packet, relayer)
+	}
+	logger := ctx.Logger().With(
+		"module", ModuleName,
+		"packet_source", packet.SourcePort,
+		"packet_destination", packet.DestinationPort,
+		"packet_sequence", packet.Sequence)
+
+	rollappPortOnHub, rollappChannelOnHub := packet.DestinationPort, packet.DestinationChannel
+	rollappID, transferPacketData, err := im.delayedAckKeeper.ExtractRollappIDAndTransferPacket(ctx, packet, rollappPortOnHub, rollappChannelOnHub)
+	if err != nil {
+		logger.Error("Failed to extract rollapp id from packet", "err", err)
+		return channeltypes.NewErrorAcknowledgement(err)
+	}
+
+	if rollappID == "" {
+		logger.Debug("Skipping IBC transfer OnRecvPacket for non-rollapp chain")
+		return im.IBCModule.OnRecvPacket(ctx, packet, relayer)
+	}
+
+	// parse the transfer amount
+	transferAmount, ok := sdk.NewIntFromString(transferPacketData.Amount)
+	if !ok {
+		err = errors.New("unable to parse transfer amount into math.Int")
+		return channeltypes.NewErrorAcknowledgement(err)
+	}
+
+	// get fee
+	feePercentage := im.GetBridgingFee(ctx)
+	fee := feePercentage.MulInt(transferAmount).TruncateInt()
+
+	// update packet data for the fee charge
+	feePacket := *transferPacketData
+	feePacket.Amount = fee.String()
+	feePacket.Receiver = im.GetFeeRecipient(ctx).String()
+
+	// No event emitted, as we called the transfer keeper directly (vs the transfer middleware)
+	err = im.transferKeeper.OnRecvPacket(ctx, packet, feePacket)
+	if err != nil {
+		logger.Error("Failed to charge bridging fee", "err", err)
+		// we continue as we don't want the fee charge to fail the transfer in any case
+		fee = sdk.ZeroInt()
+	} else {
+		logger.Debug("Charged bridging fee", "fee", fee)
+		ctx.EventManager().EmitEvent(
+			sdk.NewEvent(
+				EventTypeBridgingFee,
+				sdk.NewAttribute(AttributeKeyFee, fee.String()),
+				sdk.NewAttribute(sdk.AttributeKeySender, transferPacketData.Sender),
+			),
+		)
+	}
+
+	// transfer the remaining amount
+	transferPacketData.Amount = transferAmount.Sub(fee).String()
+	packet.Data = transferPacketData.GetBytes()
+	return im.IBCModule.OnRecvPacket(ctx, packet, relayer)
```

Comment:
I guess we rely on the next bit of the stack to catch any negative errors?
if so, comment

Reply:
@mtsitrin

Reply:
it's quite obvious by itself. used the same way in other middlewares